ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -99,79 +107,61 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
 
     /* New can fail; will deal with this later. */
     fs->connection = new S3Connection(awsConfig);
-
+    fs->fileSystem.fs_directory_list = S3ObjectList;
+    fs->fileSystem.fs_directory_list_single = S3ObjectListSingle;
+    fs->fileSystem.fs_directory_list_free = S3ObjectListFree;
     fs->fileSystem.terminate = S3FileSystemTerminate;
 
     /* TODO: Move these into tests. Just testing here temporarily to show all functions work. */
     {
-        /* List S3 buckets. */
         std::vector<std::string> buckets;
-        if (fs->connection->ListBuckets(buckets)) {
-            std::cout << "All buckets under my account:" << std::endl;
-            for (const std::string &bucket : buckets) {
-                std::cout << "  * " << bucket << std::endl;
-            }
-            std::cout << std::endl;
-        }
+        fs->connection->ListBuckets(buckets);
+        std::cout << "All buckets under my account:" << std::endl;
+        for (const std::string &bucket : buckets)
+            std::cout << "  * " << bucket << std::endl;
+        std::cout << std::endl;
 
         /* Have at least one bucket to use. */
         if (!buckets.empty()) {
-            const Aws::String firstBucket = buckets.at(0);
-
-            /* List objects. */
-            std::vector<std::string> bucketObjects;
-            if (fs->connection->ListObjects(firstBucket, bucketObjects)) {
-                std::cout << "Objects in bucket '" << firstBucket << "':" << std::endl;
-                if (!bucketObjects.empty()) {
-                    for (const auto &object : bucketObjects) {
-                        std::cout << "  * " << object << std::endl;
-                    }
-                } else {
-                    std::cout << "No objects in bucket." << std::endl;
-                }
-                std::cout << std::endl;
-            }
+            const std::string firstBucket = buckets.at(0);
 
             /* Put object. */
             fs->connection->PutObject(firstBucket, "WiredTiger.turtle", "WiredTiger.turtle");
 
-            /* List objects again. */
-            bucketObjects.clear();
-            if (fs->connection->ListObjects(firstBucket, bucketObjects)) {
-                std::cout << "Objects in bucket '" << firstBucket << "':" << std::endl;
-                if (!bucketObjects.empty()) {
-                    for (const auto &object : bucketObjects) {
-                        std::cout << "  * " << object << std::endl;
-                    }
-                } else {
-                    std::cout << "No objects in bucket." << std::endl;
-                }
-                std::cout << std::endl;
+            /* Testing directory list. */
+            WT_SESSION *session = NULL;
+            const char *prefix = "WiredTiger";
+            char ***objectList = (char ***)malloc(sizeof(char **));
+            uint32_t count;
```

Comment:
@sulabhM , should we check if `malloc` fails and return `ENOMEM`? We do this check in local_store.c

Reply:
I'm a bit confused here. I think no need to malloc a variable. Just declare a local variable char **objectList and pass &objectList to the following functions.

Reply:
You're right, that sounds good - thanks @quchenhao!